packages/sirius-web/backend/sirius-web-infrastructure/src/main/java/org/eclipse/sirius/web/infrastructure/elasticsearch/services/ProjectIndexLifecycleManager.java

```diff
@@ -0,0 +1,98 @@
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *     Obeo - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.sirius.web.infrastructure.elasticsearch.services;
+
+import java.io.IOException;
+import java.util.Objects;
+import java.util.Optional;
+
+import org.eclipse.sirius.components.core.api.IEditingContext;
+import org.eclipse.sirius.components.core.api.IEditingContextSearchService;
+import org.eclipse.sirius.web.domain.boundedcontexts.project.events.ProjectCreatedEvent;
+import org.eclipse.sirius.web.domain.boundedcontexts.project.events.ProjectDeletedEvent;
+import org.eclipse.sirius.web.domain.boundedcontexts.semanticdata.events.SemanticDataUpdatedEvent;
+import org.eclipse.sirius.web.infrastructure.elasticsearch.services.api.IEditingContextIndexingService;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.scheduling.annotation.Async;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.event.TransactionalEventListener;
+
+import co.elastic.clients.elasticsearch.ElasticsearchClient;
+
+/**
+ * Manages the lifecycle of project indices.
+ *
+ * @author gdaniel
+ */
+@Service
+public class ProjectIndexLifecycleManager {
+
+    public static final String PROJECT_INDEX_NAME_PREFIX = "project-";
+
+    private final Optional<ElasticsearchClient> optionalElasticSearchClient;
+
+    private final IEditingContextSearchService editingContextSearchService;
+
+    private final IEditingContextIndexingService editingContextIndexer;
+
+    private final Logger logger = LoggerFactory.getLogger(ProjectIndexLifecycleManager.class);
+
+    public ProjectIndexLifecycleManager(Optional<ElasticsearchClient> optionalElasticSearchClient, IEditingContextSearchService editingContextSearchService,
+            IEditingContextIndexingService editingContextIndexer) {
+        this.optionalElasticSearchClient = Objects.requireNonNull(optionalElasticSearchClient);
+        this.editingContextSearchService = Objects.requireNonNull(editingContextSearchService);
+        this.editingContextIndexer = Objects.requireNonNull(editingContextIndexer);
+    }
+
+    @Async
+    @TransactionalEventListener
+    public void onProjectCreatedEvent(ProjectCreatedEvent projectCreatedEvent) {
+        this.optionalElasticSearchClient.ifPresent(elasticSearchClient -> {
+            String createdProjectId = projectCreatedEvent.project().getId();
+            try {
+                if (!elasticSearchClient.indices().exists(existsRequest -> existsRequest.index(PROJECT_INDEX_NAME_PREFIX + createdProjectId)).value()) {
+                    elasticSearchClient.indices().create(createIndexRequest -> createIndexRequest
+                            .index(PROJECT_INDEX_NAME_PREFIX + createdProjectId)
+                            .settings(settingsBuilder -> settingsBuilder.mode("lookup"))
```

Comment:
Lookup indices are required to support joins.
They are mostly regular indices, excepted that they are only one shard, meaning that they can only contain 2 billion elements.
More information in the [documentation](https://www.elastic.co/blog/esql-lookup-join-elasticsearch)